src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMObjectWriter.cs

```diff
@@ -168,63 +174,122 @@ private unsafe void EmitObjectNode(ObjectNode node, ObjectData nodeContents)
-            dataSymbol.Section = node.GetSection(_nodeFactory).Name;
-            dataSymbol.Alignment = (uint)nodeContents.Alignment;
+            // If we need to create unaligned relocs then use a struct for the symbol.
+            // Start with the premise that most nodes will just contain aligned relocs.
+            bool useStruct = false;
+
+            if (_currentObjectData.Length < dataSizeInElements)
+            {
+                Array.Resize(ref _currentObjectData, Math.Max(_currentObjectData.Length * 2, dataSizeInElements));
+            }
+
+            Span<LLVMValueRef> dataElements = _currentObjectData.AsSpan(0, dataSizeInElements);
+            ReadOnlySpan<byte> data = nodeContents.Data.AsSpan();
+
+            // Indicies in byte units to allow us to "zip" the binary data with the relocs
+            int dataOffset = 0;
+            int relocIndex = 0;
+            int elementOffset = 0;
 
-            // Emit the value of this object node.
-            if (_currentObjectData.Length < dataSizeInPointers)
+            int relocLength = nodeContents.Relocs.Length;
+            int nextRelocOffset = 0;
+            bool nextRelocValid = relocIndex < relocLength;
+            if (nextRelocValid)
             {
-                Array.Resize(ref _currentObjectData, Math.Max(_currentObjectData.Length * 2, dataSizeInPointers));
+                nextRelocOffset = nodeContents.Relocs[0].Offset;
             }
-            Span<LLVMValueRef> dataElements = _currentObjectData.AsSpan(0, dataSizeInPointers);
-            dataElements.Clear();
 
-            // Emit relocations. We assume these are always aligned.
-            foreach (Relocation reloc in nodeContents.Relocs)
+            while (dataOffset < dataSizeInBytes)
             {
-                long delta;
-                fixed (void* location = &nodeContents.Data[reloc.Offset])
+                if (!useStruct && nextRelocValid && nextRelocOffset % pointerSize != 0)
                 {
-                    delta = Relocation.ReadValue(reloc.RelocType, location);
+                    // Switch from array to struct.
+                    useStruct = true;
+                    dataSizeInElements = nodeContents.Relocs.Length + dataSizeInBytes - (nodeContents.Relocs.Length * pointerSize);
+
+                    if (_currentObjectData.Length < dataSizeInElements)
+                    {
+                        Array.Resize(ref _currentObjectData, Math.Max(_currentObjectData.Length * 2, dataSizeInElements));
+                        Array.Resize(ref _currentObjectTypes, Math.Max(_currentObjectTypes.Length * 2, dataSizeInElements));
+                    }
+
+                    dataElements = _currentObjectData.AsSpan(0, dataSizeInElements);
+                    typeElements = _currentObjectTypes.AsSpan(0, dataSizeInElements);
+
+                    // Restart zipping while loop.
+                    dataOffset = elementOffset = relocIndex = 0;
```

Comment:
Nit: I think it's a little unconventional for runtime code style to use assignment expressions.